zict/async_buffer.py

```diff
@@ -0,0 +1,157 @@
+from __future__ import annotations
+
+import asyncio
+import contextvars
+from collections.abc import Callable, Collection
+from concurrent.futures import ThreadPoolExecutor
+from functools import wraps
+from typing import Any, Literal
+
+from zict.buffer import Buffer
+from zict.common import KT, VT, T, locked
+
+
+class AsyncBuffer(Buffer[KT, VT]):
+    """Extension of :class:`~zict.Buffer` that allows offloading all reads and writes
+    from/to slow to a separate worker thread.
+
+    This requires ``fast`` to be fully thread-safe (e.g. a plain dict).
+
+    ``slow.__setitem__`` and ``slow.__getitem__`` will be called from the offloaded
+    thread, while all of its other methods (including, notably for the purpose of
+    thread-safety consideration, ``__contains__`` and ``__delitem__``) will be called
+    from the main thread.
+    """
+
+    executor: ThreadPoolExecutor | None
+    futures: set[asyncio.Future]
+    evicting: float | None
+
+    @wraps(Buffer.__init__)
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self.executor = None
+        self.futures = set()
+        self.evicting = None
+
+    def close(self) -> None:
+        # Call LRU.close(), which stops LRU.evict_until_below_target() halfway through
+        super().close()
+        if self.executor is not None:
+            # TODO just use shutdown(cancel_futures=True) (requires Python >=3.9)
+            for future in self.futures:
+                future.cancel()
+            self.executor.shutdown(wait=True)
+            self.executor = None
+
+    def _offload(self, func: Callable[..., T], *args: Any) -> asyncio.Future[T]:
+        if self.executor is None:
+            self.executor = ThreadPoolExecutor(
+                1, thread_name_prefix="zict.Buffer offloader"
+            )
+
+        loop = asyncio.get_running_loop()
+        context = contextvars.copy_context()
+        future = loop.run_in_executor(self.executor, context.run, func, *args)
+        self.futures.add(future)
+        future.add_done_callback(self.futures.remove)
+        return future  # type: ignore[return-value]
+
+    # Return an asyncio.Future, instead of just writing it as an async function, to make
+    # it easier for overriders to tell apart the use case when all keys were already
+    # in fast
+    @locked
+    def async_get(
+        self, keys: Collection[KT], missing: Literal["raise", "omit"] = "raise"
+    ) -> asyncio.Future[dict[KT, VT]]:
```

Comment:
Mostly curious, I don't think we should change this now but why is the API accepting multiple keys? is this somehow a requirement or an optimization?

The reason why I'm asking is because I _think_ this drives some complexity and I would expect `missing` not to be required otherwise. I'm not entirely sure about this.

Reply:
It's an optimization to avoid having an extremely long queue to the thread pool, with each individual key incurring in thread synchronization overhead.